shared/shared.go

```diff
@@ -75,26 +75,38 @@ func MakeLabelFunc() func(hash LabelHash, labelID uint64, leftSiblings [][]byte)
 	}
 }
 
+type Member struct {
+	Challenge []byte `scale:"max=32"`
+}
+
 type Proof struct {
 	// The actual proof.
 	MerkleProof
 
 	// Members is the ordered list of miners challenges which are included
 	// in the proof (by using the list hash digest as the proof generation input (the statement)).
-	Members [][]byte
+	Members []Member `scale:"max=150"`
 
 	// NumLeaves is the width of the proof-generation tree.
 	NumLeaves uint64
 }
 
 type ProofMessage struct {
 	Proof
-	ServicePubKey []byte
-	RoundID       string
+	ServicePubKey []byte `scale:"max=32"`
+	RoundID       string `scale:"max=32"`
+}
+
+type Leaf struct {
+	Value []byte `scale:"max=32"`
+}
+
+type Node struct {
+	Value []byte `scale:"max=32"`
 }
 
 type MerkleProof struct {
-	Root         []byte
-	ProvenLeaves [][]byte
-	ProofNodes   [][]byte
+	Root         []byte `scale:"max=32"`
+	ProvenLeaves []Leaf `scale:"max=150"`
+	ProofNodes   []Node `scale:"max=5400"` // 36 nodes per leaf
```

Comment:
Why can't it stay `[][]byte`? These new types generated a lot of additional boilerplate code.

Reply:
Because there is no simple way to specify the size of recursive slices:

```
type MyStruct struct {
    Field [][]byte `scale:"max=10"` // only specifies that len(Field) <= 10, not that len(Field[0]) <= 10
}

Reply:
But it complicates the existing code _a lot_. I mean code just converting from `[]byte` to the new structs, like:
```go
pLeaves := make([]shared.Leaf, 0, len(provenLeaves))
for _, leaf := range provenLeaves {
	pLeaves = append(pLeaves, shared.Leaf{
		Value: leaf,
	})
}
pNodes := make([]shared.Node, 0, len(proofNodes))
for _, node := range proofNodes {
	pNodes = append(pNodes, shared.Node{
		Value: node,
	})
}
```

How about doing it this way:
1. make a copy of the existing structures (the original ones) - "not-shared", use them inside of this repo - avoiding all the new boilerplate
2. keep the new structures (with added scale tags) in `package shared` for go-spacemesh to use
3. In follow up PRs move the structures from `shared` directly to go-spacemesh repo

It will limit the added boilerplate, allow you to quickly proceed with this PR, and clean the code up soon.

Reply:
Which tests call `prover.GenerateProof`? I can't find them.

I looked through the code and as far as I can see, go-spacemesh (de)serializes `shared.MerkleProof` only. `shared.Proof` is (de)serialized in poet for storage in DB, scale was used out of convenience only - any serialization could be used, for example, `xdr` which is also already used in poet.

Bonus: Using something like `xdr` for Proof to save in DB would also remove the problem of deciding about the maximum number of members in Proof.

>I don't think this code is convoluted or complicated, its a simple conversion from one type to another.

It obscures the real logic in code w/o a good reason, or just because some application that depends on the poet's verification method also happens to serialize the Merkle proof for some reasons (reasons that the poet doesn't care about).

IMO the problem is that a requirement of go-spacemesh (that is - to be able to (de)serialize `shared.MerkleProof` leaks into the poet library).

Reply:
> Not really. If the members slice can be arbitrarily large we need to store an arbitrarily large object in our DB. We should limit the size independent of the encoder we use.
> 
> I would even go so far and say that if the maximum amount of users have submitted their challenges to poet, it should just refuse any additional request. We will have scaling issues if every user submits their challenge to every poet and every poet is supposed to store and process all of that data.

That is exactly how this should be enforced - on the API level, not during serialization. Otherwise, the poet will run for the entire epoch and fail miserably trying to save the proof in DB. It would be really bad.

I quickly refactored the existing code to not rely on scale for other structs than `shared.MerkleProof`: #272. Merging it first would limit footprint of this PR.